cli/fathom_web/commands/label.py

```diff
@@ -0,0 +1,112 @@
+                               f' against unwanted data loss, please move or remove the existing directory.')
+    else:
+        originals_dir = None
+
+    list_of_items = os.listdir(in_directory)
+    number_of_items = len(list_of_items)
+
+    print_statements = [] # Capture any print statements to log at the end.
+
+    # Make a pool of workers. Each worker is in its own process. We use a scaling factor to account for the overhead of
+    # setting up all of the processes.
+    pool = multiprocessing.Pool(multiprocessing.cpu_count() * 4)
+    # Curry ``task``, so we can pass more than one argument into pool.imap_unordered.
+    task = partial(label_task, in_directory, in_type, originals_dir, preserve_originals)
+
+    with progressbar(pool.imap_unordered(task, list_of_items),
+                    label='Labeling pages',
+                    length = len(list_of_items)) as bar:
+        for result in bar:
+            if result is not None:
+                print_statements.append(result)
+            pass
+
+    for statement in print_statements:
+        print(statement)
+
+
+def label_task(in_directory, in_type, originals_dir, preserve_originals, filename):
+    file = pathlib.Path(in_directory) / filename
+    if file == originals_dir:
+        return
+    if file.is_dir():
+        return f'Skipped directory {file.name}/'
+    if file.suffix != '.html':
+        return f'Skipped {file.name}; not an HTML file'
+
+    with file.open(encoding='utf-8') as fp:
+        html = fp.read()
+
+    new_html = label_html_tags_in_html_string(html, in_type)
+
+    if preserve_originals:
+        shutil.move(file, originals_dir / file.name)
+
+    with file.open('w', encoding='utf-8') as fp:
+        fp.write(new_html)
+
+
+class HTMLParserSubclass(HTMLParser):
+    def __init__(self, in_type, **kwargs):
+        self.in_type = in_type
+        self.html_tags_list = []
+        super().__init__(**kwargs)
+
+    def handle_starttag(self, tag, attrs):
+        if tag == 'html':
+            original_html_tag = self.get_starttag_text()
+            new_html_substring = f'html data-fathom="{self.in_type}"'
+            new_html_tag = original_html_tag.replace('html', new_html_substring, 1)
+            self.html_tags_list.append(tuple([original_html_tag, new_html_tag]))
+
+
+def label_html_tags_in_html_string(html: str, in_type: str) -> str:
+    """
+    Finds all opening ``html`` tags in the HTML string and adds a
+    ``' data-fathom="${in_type}"'`` substring to each one.
+
+    We do this by building a new HTML string with the inserted substring(s).
+
+    The ``html`` tags are found using the HTMLParser class in Python's
+    built-in html.parser library.
+    """
+    parser = HTMLParserSubclass(in_type)
+    parser.feed(html)
+
+    new_html = html
+
+    for (original_html_tag, new_html_tag) in parser.html_tags_list:
+        new_html = new_html.replace(original_html_tag, new_html_tag, 1)
```

Comment:
This actually makes us do the work of scanning through the string twice: once in the parser, and again here, to find the previously found string and replace it. Did my suggestion of HTMLParser.getpos() not work out?

Reply:
I couldn't figure out how to do it with `getpos()`. Inside `handle_starttag`, if I call `self.getpos()`, it returns `(2, 0)` -- i.e. line 2, with offset of 0 -- i.e. the line number and offset for the opening `<html>` tag.

Maybe I am misunderstanding how to use [`getpos()`](https://docs.python.org/3/library/html.parser.html#html.parser.HTMLParser.getpos), but I don't see how it can be used to give me the offset within the opening `<html>` tag for where to insert the new `data-fathom` attribute.

Reply:
Even if I could get that, I don't see a method in the class to replace the opening `<html>` tag with a new value.

Reply:
Alright, that means this wasn't the best choice of parser for this job. But it's good enough for research code; it's not like we're running this in tight loops.